[MIG] account_global_discount: 16.0

ferran-S73 opened the pull request:
Depends on:
- [ ] https://github.com/OCA/server-backend/pull/221

~~Hi I'm gonna need some help migrating this module @OCA/accounting-maintainers @kirca @omar7r @pedrobaeza~~

~~[Many accounting code has been refactored in v16](https://github.com/odoo/odoo/commit/d8d47f9ff8554f4b39487fd2f13c153c7d6f958d) and so my approach here, since there's no `_recompute_tax_lines` nor `_recompute_dynamic_lines` is to do this module's calculations whenever an invoice is created or the field `global_discount_ids` is written following indications from Odoo's commit description.~~

~~I'm not quite getting the desired results and I was wondering if you could help me pointing if this is a viable approach or if you have any other idea. 
I also looked at [this migration](https://github.com/OCA/account-invoicing/pull/1269/commits/e628d2a778179d36636af6b40c35a10b93159bc4) but can't seem to figure how could I do something similar here.~~

~~Thanks in advance to anyone who can help~~



In order to adapt this module to the new way of handling invoices in 16.0 I moved the creation / update of discount lines to the `create` and `write` methods of the `account.move`. We no longer need the recalc the bas amount for the discount lines since it is done automatically by Odoo as well as the total amounts of the invoice.

Also, for the field `global_discount_item` which was needed because `invoice_global_discount_id` wasn't properly filled, I checked and that last field is now handled correctly. I'm not removing the `global_discount_item` altogether for compatibility reasons with old invoices since there's no clear way to migrate it to fill `invoice_global_discount_id`.

chienandalu committed:
[ADD][11.0] account_global_discount: New Module

chienandalu committed:
[FIX+IMP] account_global_discount: link line taxes to discount move line + Discounts in invoice report

[IMP] account_global_discount: Discounts in invoice report

Give discounts info in the invoice report.

[FIX] account_global_discount: link line taxes to discount move line

pedrobaeza committed:
[FIX] account_global_discount: More on global discount/taxes link

- We need to take into account invoice lines with multiple taxes, so the link should
  be m2m.
- Migration scripts for preserving the best information on existing installations.
- Tests for checking new conditions.
- Perform sanity check for not ending in an incompatible situation.
- Some refactor done on onchanges for avoiding duplicating operations.
- Adjust UI for not allowing to edit computed invoice global discounts.

pedrobaeza committed:
[MIG] acount_global_discount: Migration to 12.0

- Standard procedure
- Make test more resilient

pedrobaeza committed:
[IMP] account_global_discount: Deprecated use of currency compute

Replaced by more explicit `_convert` method.

pedrocs-exo committed:
Added translation using Weblate (Portuguese)

pedrobaeza committed:
[FIX] account_global_discount: Fix singleton error in _compute_amount

For reducing diff, we rename the method and call it record per record.

Fixes #788

pedrocs-exo committed:
[FIX][12.0] account_global_discount: sale order templates compatibility

Commit:
[IMP] account_global_discount: black, isort, prettier

Commit:
[MIG] account_global_discount: Migration to 13.0

[UPD] Update account_global_discount.pot

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-13.0/account-invoicing-13.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_global_discount/

chienandalu committed:
[IMP] account_global_discount: security group

Users will need a specific group to set global discounts so it's easier
to decide who's allowed to apply them.

[UPD] Update account_global_discount.pot

Commit:
[FIX] account_global_discount: wrong tax move line discounts on invoice post

OCA-git-bot committed:
account_global_discount 13.0.1.0.1

victoralmau committed:
[FIX] account_global_discount: Compute global discount with 0% taxes

Co-Authored-By: Pedro M. Baeza <pedro.baeza@tecnativa.com>

[UPD] README.rst

Commit:
Added translation using Weblate (Dutch)

Commit:
Translated using Weblate (Dutch)

Currently translated at 100.0% (35 of 35 strings)

Translation: account-invoicing-13.0/account-invoicing-13.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_global_discount/nl_NL/

[UPD] Update account_global_discount.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-13.0/account-invoicing-13.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_global_discount/

anasuarez1 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (34 of 34 strings)

Translation: account-invoicing-13.0/account-invoicing-13.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_global_discount/es/

victoralmau committed:
[FIX] account_global_discount: Fix tests

OCA-git-bot committed:
account_global_discount 13.0.1.0.2

chienandalu committed:
[FIX] account_global_discount: discounts on create

When we set the global discount on the create method we must ensure that
the move lines are properly recomputed

OCA-git-bot committed:
account_global_discount 13.0.1.1.0

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-13.0/account-invoicing-13.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_global_discount/

victoralmau committed:
[FIX] account_global_discount: Prevent error when add more taxes in multi-lines from invoices when global discount is empty

OCA-git-bot committed:
account_global_discount 13.0.1.1.1

pedrobaeza committed:
[FIX+REF] account_global_discount: Include taxes in global discount move lines

Previously, no taxes were populated as base taxes for global discount
move lines, which means that the tax reports were incorrect.

The global discount lines + move lines has been injected other way for
avoiding inconsistencies, and the rest of the features have been adapted
according this.

A migration script is provided as well for filling taxes in existing
global discount move lines.

[UPD] Update account_global_discount.pot

[UPD] README.rst

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-13.0/account-invoicing-13.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_global_discount/

pedrobaeza committed:
[FIX] account_global_discount: Call the proper method on create

For recomputing the global discounts, now we need to call another method.

TT30077

omar7r committed:
[MIG] account_global_discount: Migration to 14.0

omar7r committed:
[IMP] account_global_discount: securization for multi-company

omar7r committed:
[FIX] account_global_discount: Signed amount error in supplier invoices

omar7r committed:
[FIX] account_global_discount: computed fields must be stored because the other standard computed fields in same method are

omar7r committed:
[FIX] account_global_discount: Filtering global discount in on_change by invoice's company

omar7r committed:
[IMP] account_global_discount: Global discounts into invoice analysis report

omar7r committed:
[FIX] account_global_discount: Partner in global discount move line

omar7r committed:
[FIX] account_global_discount: When writing Odoo doesn't autocomplete amount_currency and it could fail with constraint

oca-travis committed:
[UPD] Update account_global_discount.pot

OCA-git-bot committed:
[UPD] README.rst

francesco-ooops committed:
Added translation using Weblate (Italian)

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (36 of 36 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_global_discount/it/

pedrobaeza committed:
[FIX] *: Change _recompute_tax_lines args according upstream

https://github.com/odoo/odoo/commit/c9d5f069739d897145f843ee93b7eefd1d39
has changed the number of arguments on the method `_recompute_tax_lines`,
so we need to adapt the overrides to such change.

OCA-git-bot committed:
account_global_discount 14.0.1.0.1

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 100.0% (35 of 35 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_global_discount/pt/

kirca committed:
[MIG] account_global_discount: Migration to 15.0

kirca committed:
[FIX] account_global_discount: multi-currency

In a foreign currency invoice the amounts (amount_total,
amount_untaxed) are expressed in the `invoice` currency while the
accounting entry lines are expressed in the `company` currency.

The base_before_global_discounts has to be expressed in the invoice
currency as well in such invoices because the discounts need to be
computed in the same currency as the other amounts.

Also set the invoice currency_id on the tax line so that it doesn't
default to the company currency, and convert the discount journal entry
amount to the company currnecy.

ferran-S73 committed:
[IMP] account_global_discount: black, isort, prettier

ferran-S73 committed:
[MIG] account_global_discount: 16.0

pedrobaeza commented:
I can't say too much. Maybe you have to override `_inverse_tax_totals`.

ferran-S73 commented:
Hey @OCA/accounting-maintainers this should be ready for review now! :partying_face:

ferran-S73 committed:
[DON'T MERGE] test-requirements.txt

tafaRU commented:
/ocabot migration account_global_discount

ferran-S73 commented:
Hi @OCA/accounting-maintainers I'm having trouble migrating the tests for this module, for some reason when a global discount is added a new tax line is added too instead of updating the existing one. Does anyone have encountered a similar problem? I'd be very thankful if anyone could help me out here. Thanks in advance

BT-srieskamp commented:
@ferran-S73 not sure if this is the right place for feedback to that module. I tested everything and it looks quite good from a users perspective. But the two new lines at the bottom of the Sale Order and Invoice are not in the right place. Maybe this is already fixed? Can you reproduce this? I added an attachment.
![Screenshot 2023-04-25 at 14 01 45](https://user-images.githubusercontent.com/59617476/234270710-716d9b08-4187-46fa-b31b-29172abfd691.png)

ferran-S73 commented:
> @ferran-S73 not sure if this is the right place for feedback to that module. I tested everything and it looks quite good from a users perspective. But the two new lines at the bottom of the Sale Order and Invoice are not in the right place. Maybe this is already fixed? Can you reproduce this? I added an attachment. ![Screenshot 2023-04-25 at 14 01 45](https://user-images.githubusercontent.com/59617476/234270710-716d9b08-4187-46fa-b31b-29172abfd691.png)

Hi I'll try to reproduce with the same currency and reach you back with anything I find out @BT-skettler

ChrisOForgeFlow committed:
[14.0][IMP] account_global_discount, added init hook to avoid memory issues on big database installation

Rferri44-S73 committed:
[16.0][IMP] account_global_discount: Remake amount global discount

ferran-S73 commented:
@OCA/accounting-maintainers

rafaelbn commented:
/ocabot rebase

rafaelbn commented:
Hello @ferran-S73 ,

Is this whay you except?

![imagen](https://github.com/OCA/account-invoicing/assets/8736623/2b6c6a90-5248-459d-add9-ba3a7790972a)


I have beed testing and I don't like to much this module, complicated. REAMDE in configuration should be reviewed.

Any case, testing is OK no error, just some maybe UX?

Up to you!

rafaelbn reviewed (approved):
Functional basic test in runbot 👍🏼

rafaelbn commented:
You can review here @BT-skettler ! 🥳

BT-srieskamp commented:
> 

@ferran-S73 I hope I tested it correctly. On runbot I could see that the summary below the sale order line is aligned (with currency CHF).
Thanks a lot for you effort!

OCA-git-bot committed:
[FIX] account_global_discount: Make tests resilient

- Depending on the installed set of modules, the company currency may
  be USD or EUR. If the second case, these tests will fail, so we make
  sure that the company currency is USD for our tests, doing the change
  by SQL, as there's a Python constraint that prevents it.

Not needed in v17 due to: odoo/odoo#107113.

victoralmau commented:
Please, cherry-pick https://github.com/OCA/account-invoicing/pull/1531 to commit history

ferran-S73 commented:
> Please, cherry-pick #1531 to commit history

@victoralmau done

QuocDuong1306 reviewed (approved):
Tested, lgtm

cyrilmanuel reviewed (approved):
not an expert but lgtm

rafaelbn commented:
Thank you @ferran-S73 ! @victoralmau your comments are attended, are you agree? 😄

linuxivan commented:
There is an error when returning an invoice, it re-applies the discount on the already discounted invoice and the amount is not the same, resulting in a partial payment on the original invoice.
![image](https://github.com/OCA/account-invoicing/assets/61226403/382e7c6d-412b-4450-85b4-b668ca351699)

JavierIniesta reviewed (approved):
Code Review: LGTM

Could you add unit tests to cover the missing use cases?

linuxivan commented:
> Code Review: LGTM
> 
> Could you add unit tests to cover the missing use cases?

You have to look better.

hildickethan commented:
> There is an error when returning an invoice, it re-applies the discount on the already discounted invoice and the amount is not the same, resulting in a partial payment on the original invoice.

Fixed, now when creating a refund it will clear the global discount line and recalc it
Also fixed a bug with the "apply discount" field not working correctly

cyrilmanuel reviewed (approved):
not an expert but lgtm

simahawk commented:
FTR still blocked by its dependency https://github.com/OCA/server-backend/pull/221

PauBForgeFlow commented:
Hi, I think it would be interesting to cherry-pick this [#PR1448 ](https://github.com/OCA/account-invoicing/pull/1448) to commit history.

miguel-S73 commented:
@PauBForgeFlow `cherry-pick` added.

carolinafernandez-tecnativa reviewed (approved):
Functional review LGTM :) Thanks!
TT45863

rafaelbn commented:
- [ ] Depends in https://github.com/OCA/server-backend/pull/221

GasparJarsa commented on test-requirements.txt:
```diff
@@ -0,0 +1 @@
+odoo-addon-base_global_discount @ git+https://github.com/OCA/server-backend.git@refs/pull/221/head#subdirectory=setup/base_global_discount
```
This module is already merged, could you please remove this file in order to pass the tests in Github Actions.

carolinafernandez-tecnativa commented:
Is there anything else to do? so we can unlock some migrations!
Thanks!

ping @pedrobaeza

pedrobaeza commented:
For unlocking the situation, I have extracted the migration from this PR and remove the depending change in #1724 

Closing this one. When the other PR gets unlocked, you can propose the improvement in this module in a new PR.
